soroban-sdk/src/deploy.rs

```diff
@@ -53,61 +56,102 @@ impl Deployer {
 
-    /// Get a deployer that deploys contracts that derive their contract IDs
-    /// from the current contract and the provided salt.
+    /// Get a deployer that deploys contract that derive the contract IDs
+    /// from the current contract and provided salt.
     pub fn with_current_contract(
         &self,
-        salt: &impl IntoVal<Env, BytesN<32>>,
-    ) -> DeployerWithCurrentContract {
-        let env = self.env();
-        DeployerWithCurrentContract {
-            env: env.clone(),
-            salt: salt.into_val(env),
+        salt: impl IntoVal<Env, BytesN<32>>,
+    ) -> DeployerWithAddress {
+        DeployerWithAddress {
+            env: self.env.clone(),
+            address: self.env.current_contract_address(),
+            salt: salt.into_val(&self.env),
         }
     }
 
-    #[doc(hidden)]
-    /// Get a deployer for contracts that derive their contract IDs from the
-    /// given contract ID and the provided salt.
-    pub fn with_other_contract(
+    /// Get a deployer that deploys contracts that derive the contract ID
+    /// from the provided address and salt.
+    ///
+    /// The deployer address must authorize all the deployments.
+    pub fn with_address(
+        &self,
+        address: Address,
+        salt: impl IntoVal<Env, BytesN<32>>,
+    ) -> DeployerWithAddress {
+        DeployerWithAddress {
+            env: self.env.clone(),
+            address,
+            salt: salt.into_val(&self.env),
+        }
+    }
+
+    /// Get a deployer that deploys an instance of Stellar Asset Contract
+    /// corresponding to the provided serialized asset.
+    ///
+    /// `serialized_asset` is the Stellar `Asset` XDR serialized to bytes. Refer
+    /// to `stellar-xdr` create for the exact definition.
+    pub fn with_stellar_asset(
         &self,
-        contract_id: &Address,
-        salt: &impl IntoVal<Env, BytesN<32>>,
-    ) -> DeployerWithOtherContract {
-        let env = self.env();
-        DeployerWithOtherContract {
-            env: env.clone(),
-            contract_id: contract_id.clone(),
-            salt: salt.into_val(env),
+        serialized_asset: impl IntoVal<Env, Bytes>,
+    ) -> DeployerWithAsset {
+        DeployerWithAsset {
+            env: self.env.clone(),
+            serialized_asset: serialized_asset.into_val(&self.env),
         }
     }
+
+    /// Upload the contract Wasm code to the network.
+    ///
+    /// Returns the hash of the uploaded Wasm that can be then used for
+    /// the contract deployment.
+    /// ### Examples
+    /// ```
+    /// use soroban_sdk::{BytesN, Env};
+    ///
+    /// const WASM: &[u8] = include_bytes!("../doctest_fixtures/contract.wasm");
+    ///
+    /// #[test]
+    /// fn test() {
+    /// # }
+    /// # fn main() {
+    ///     let env = Env::default();
+    ///     env.deployer().upload_contract_wasm(WASM);
+    /// }
+    /// ```
+    pub fn upload_contract_wasm(&self, contract_wasm: impl IntoVal<Env, Bytes>) -> BytesN<32> {
```

Comment:
@dmkozh I thought for some reason you had advocated previously for upload being outside the Deployer type, because it isn't technically "deployment" in the contract instance case.

It may be a little odd that `update_current_contract_wasm` lives on `Env`, and yet `upload` lives in Deployer.

Should we move update here too?

Reply:
I don't really have any strong opinion on that. Previously I didn't want to allow uploading wasm from contracts, but then I figured that makes sense for uploading external wasm (and for the sake of feature parity with operation). For the update I guess I just wanted to get something working quickly. Move to deployer for now.